toolchain/check/check.cpp

```diff
@@ -30,20 +30,61 @@ class SemIRLocationTranslator
     : public DiagnosticLocationTranslator<SemIRLocation> {
  public:
   explicit SemIRLocationTranslator(
-      Parse::NodeLocationTranslator* node_translator, const SemIR::File* sem_ir)
-      : node_translator_(node_translator), sem_ir_(sem_ir) {}
+      const llvm::DenseMap<const SemIR::File*, Parse::NodeLocationTranslator*>*
+          node_translators,
+      const SemIR::File* sem_ir)
+      : node_translators_(node_translators), sem_ir_(sem_ir) {}
 
   auto GetLocation(SemIRLocation loc) -> DiagnosticLocation override {
-    if (loc.is_inst_id) {
-      return node_translator_->GetLocation(
-          sem_ir_->insts().GetParseNode(loc.inst_id));
-    } else {
-      return node_translator_->GetLocation(loc.node_location);
+    // Parse nodes always refer to the current IR.
+    if (!loc.is_inst_id) {
+      return GetLocationInFile(sem_ir_, loc.node_location);
+    }
+
+    const auto* cursor_ir = sem_ir_;
+    auto cursor_inst_id = loc.inst_id;
+    while (true) {
+      // If the parse node is valid, use it for the location.
+      if (auto parse_node = cursor_ir->insts().GetParseNode(cursor_inst_id);
+          parse_node.is_valid()) {
+        return GetLocationInFile(cursor_ir, parse_node);
+      }
+
+      // If the parse node was invalid, recurse through import references when
+      // possible.
+      if (auto import_ref = cursor_ir->insts().TryGetAs<SemIR::LazyImportRef>(
+              cursor_inst_id)) {
+        cursor_ir = cursor_ir->cross_ref_irs().Get(import_ref->ir_id);
+        cursor_inst_id = import_ref->inst_id;
+        continue;
+      }
+
+      // If a namespace has an instruction for an import, switch to looking at
+      // it.
+      if (auto ns =
+              cursor_ir->insts().TryGetAs<SemIR::Namespace>(cursor_inst_id)) {
+        if (ns->import_id.is_valid()) {
+          cursor_inst_id = ns->import_id;
+          continue;
+        }
+      }
+
+      // Invalid parse node but not an import; just nothing to point at.
+      return GetLocationInFile(cursor_ir, Parse::NodeId::Invalid);
     }
   }
 
  private:
-  Parse::NodeLocationTranslator* node_translator_;
+  auto GetLocationInFile(const SemIR::File* sem_ir,
+                         Parse::NodeLocation node_location) const
+      -> DiagnosticLocation {
+    auto it = node_translators_->find(sem_ir);
+    CARBON_CHECK(it != node_translators_->end());
+    return it->second->GetLocation(node_location);
+  }
+
+  const llvm::DenseMap<const SemIR::File*, Parse::NodeLocationTranslator*>*
```

Comment:
Minor, and fine to do in a follow-up: might be worth extracting a typedef for this map. It's used in more than one function declaration it seems in addition to here.

Reply:
I tend to prefer not doing typedefs in cases like this. https://google.github.io/styleguide/cppguide.html#Aliases I've always read as being reticent about typedefs.

Reply:
FWIW, I wasn't suggesting this really be public.

But regardless, this is optional.